### PR TITLE
feat: [IAC-750]: Kubernetes delegate support for IACM

### PR DESCRIPTION
**Pull request by scottyw-harness**
## Describe your changes

In order to support IACM stages in the Kubernetes delegate we need to make changes to the CI and IACM managers.

At plan creation time we need to include workspace, env vars and secret vars in the plan so that the containers can be correctly initialized at creation time in Kubernetes.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50182)

### CODE REVIEW

**Comment by scottyw-harness** [comment]
trigger uts

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 1]
```diff
@@ -197,6 +245,38 @@ public Map<String, SecretNGVariable> getPluginCompatibleSecretVars(PluginCompati
...
+                    .type(NGVariableType.SECRET)
+                    .value(ParameterField.createValueField(SecretRefHelper.createSecretRef(entry.getValue())))
+                    .name(entry.getKey())
+                    .build());
+          }
+        }
+        return secrets;
+      }
+      case IACM_APPROVAL: {
+        IACMApprovalInfo iacmApprovalInfo = (IACMApprovalInfo) step;
```
Comment (nassergonzalez) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:265]: I think that we could skip this for the approval step as it is just a glorified pause, and it doesn't need to have secrets injected

**Review by nassergonzalez — DISMISSED** [review]

**Review by allofthesepeople — DISMISSED** [review]

**Comment by scottyw-harness** [comment]
trigger uts

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 2]
```diff
@@ -184,8 +190,49 @@ public Map<String, String> getPluginCompatibleEnvVariables(PluginCompatibleStep
       case SSCA_ENFORCEMENT:
         return SscaEnforcementPluginHelper.getSscaEnforcementStepEnvVariables(
             (SscaEnforcementStepInfo) stepInfo, identifier, ambiance, infraType);
+      case IACM_TERRAFORM_PLUGIN: {
+        IACMTerraformPluginInfo iacmTerraformPluginInfo = (IACMTerraformPluginInfo) stepInfo;
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:194]: please move this logic to iacm util. We can just have case which will call method in utils

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java [hunk 3]
```diff
@@ -197,6 +244,22 @@ public Map<String, SecretNGVariable> getPluginCompatibleSecretVars(PluginCompati
       case SSCA_ENFORCEMENT:
         return SscaEnforcementPluginHelper.getSscaEnforcementSecretVariables(
             (SscaEnforcementStepInfo) step, identifier);
+      case IACM_TERRAFORM_PLUGIN: {
+        IACMTerraformPluginInfo iacmTerraformPluginInfo = (IACMTerraformPluginInfo) step;
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/buildstate/PluginSettingUtils.java:248]: please move this logic to iacm util. We can just have case which will call method in utils

File: 332-ci-manager/service/src/main/java/io/harness/iacm/execution/IACMStepsUtils.java [hunk 4]
```diff
@@ -122,19 +122,40 @@ private Map<String, String> getWorkspaceVariables(
     return pluginEnvs;
   }
 
+  private Map<String, String> getSecretVariables(String org, String projectId, String accountId, String workspaceID) {
+    WorkspaceVariables[] variables = getIACMWorkspaceVariables(org, projectId, accountId, workspaceID);
+    Map<String, String> secrets = new HashMap<>();
+    for (WorkspaceVariables variable : variables) {
+      switch (variable.getKind()) {
+        case "env":
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/iacm/execution/IACMStepsUtils.java:130]: can we move this string to constant

File: 332-ci-manager/service/src/main/java/io/harness/iacm/execution/IACMStepsUtils.java [hunk 5]
```diff
@@ -122,19 +122,40 @@ private Map<String, String> getWorkspaceVariables(
...
+          if (Objects.equals(variable.getValue_type(), "secret")) {
+            secrets.put(variable.getKey(), variable.getValue());
+          }
+          break;
+        case "tf":
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/iacm/execution/IACMStepsUtils.java:135]: can we move this string to constant

**Comment by scottyw-harness** [comment]
trigger uts

**Comment by scottyw-harness** [comment]
trigger uts

**Comment by scottyw-harness** [comment]
trigger ut9

**Comment by scottyw-harness** [comment]
trigger ut0

**Review by harshjain12 — APPROVED** [review]

**Pull request merged**